docs/introduction/container-images.md

```diff
@@ -0,0 +1,253 @@
+
+While the term "virtual machine" is used in different contexts and can spawn much disagreement about
+its definition, for the time being let's abstractly talk about the differences between *containers* and
+*hypervisors*. The biggest difference was already stated - **hypervisor emulation is based on emulating hardware**. With this kind of technology, we take a machine and it emulates virtual hardware, and then bring up 
+another operating system on top of that hardware. There is a nice slide to show [hypervisor based virutualization here](https://www.slideshare.net/Docker/introduction-to-docker-2017/18).
+
+Containers are about virtualizing the subsystems of the operating system itself. We take services that the operating system provides (networking, filesystem, etc.) and provide them in a way that is virtuaizable. This means that I can bring up different copies of an operating system on the same kernel. We are using features of the kernel. So you can think containers as  building on the level of the kernel, and this is different from hypervisors that build one level down, and try to emulate the hardware.
+
+ - Containers: have a **single** kernels running underneath them
+ - Hypervisors: have **multiple** kernels
+
+A cool result of this, given that you have a single kernel, is that if you fix or update that kernel, all the "guests" running off it also get patched.
+
+For memory, a hypervisor can only see pages. It doesn't know what each guest is doing. With a container, under a single kernel the single kernel can see all objects and usage inside each container to make resource decisions. For this reason the resource decisions are made much more efficient.
+
+### What is a virtual machine?
+
+Now we can tell you that a virtual machine runs on a hypervisor, and we have a hardware layer, an entire  
+entire operating system, and then libraries and applications on top of it. A VM is a very needy fellow:
+
+ - Each VM requires CPU, storage, memory, and an entire operating system
+ - More VMs == more resources!
+ - Portability is not guaranteed
+
+The need to allocate more system resources makes them a lot fatter and less portable than containers.
+For a ballpark comparison, a hypervisor image is typically gigabytes, and a container can be in the order of megabytes. You've probably heard people talk about containers and use terms like "lightweight" and "efficient." 
+I found a nice slide to show the differences [here](https://www.slideshare.net/Docker/introduction-to-docker-2017/22).
+
+### Linux Container API
+
+So if containers are so great, why did they only "become a thing" around 2014? 
+It's because the linux containers API is pretty compiicated. Containers in linux are controlled by
+a Kernel API, and I found a good summary and diagram in the talk [here](https://www.youtube.com/watch?v=YsYzMPptB-k#t=18m10s) picture at 18 minutes, 10 seconds. There is quite some (people-based) history behind 
```

Comment:
IMO, too much detail at this stage.

You are spot on by asking "why did they become a thing in 2014?" and the answer is really that Docker came along and gave 'mere mortals' an ecosystem for building, pushing, pulling and running containers that simplifies the complexity of the underlying things like cgroups and namespaces.

Reply:
I was one of those mere mortals! Oh wait, I still am.